rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -669,8 +669,11 @@ rmw_create_publisher(
     rmw_zenoh_cpp::liveliness::EntityType::Publisher,
     rmw_zenoh_cpp::liveliness::NodeInfo{
       node->context->actual_domain_id, node->namespace_, node->name, ""},
-    rmw_zenoh_cpp::liveliness::TopicInfo{rmw_publisher->topic_name,
-      publisher_data->type_support->get_name(), publisher_data->adapted_qos_profile}
+    rmw_zenoh_cpp::liveliness::TopicInfo{
+      rmw_publisher->topic_name,
+      publisher_data->type_support->get_name(),
+      type_hash_c_str,
```

Comment:
It feels to me like there are two problems with this implementation:

1.  The ownership of `type_hash_c_str` in this function is very unclear.  In particular, this function (`rmw_create_publisher`) allocates it (via calling `rosidl_stringify_type_hash`), but then it hands off ownership to `TopicInfo` to eventually free.  While that works, it is really easy for the code to change and for a memory leak to happen later.
2.  We convert the type hash to a string in this function, and then convert it back to a `rosidl_type_hash_t` in `GraphCache::get_entities_info_by_topic`.  That just seems like unnecessary work.

To fix this, I suggest two changes:
1.  Pass `publisher_data->type_hash` in here, and just have the `TopicInfo` hold onto that.
2.  Make sure to free `type_hash_c_str` in this function (probably via a `rclcpp::make_scope_exit`).  This is actually an existing memory leak, but we may as well fix it while we are looking at things here.

The same goes for all uses of the type_hash below.

Reply:
I was relying on [implicit conversion to a string](https://github.com/ros2/rmw_zenoh/blob/54c33bcd51381725ecb6e585db71679681ac08cf/rmw_zenoh_cpp/src/detail/liveliness_utils.hpp#L52) when passing `type_hash_c_str ` to the `TopicInfo` constructor. That should create a copy right? I can also explcitly pass a `std::string(type_hash_c_str)`.

> We convert the type hash to a string in this function, and then convert it back to a rosidl_type_hash_t in GraphCache::get_entities_info_by_topic. That just seems like unnecessary work.

it's only unnecessary when the topics are created in the current session. However we still need to convert the string to a `rosidl_type_hash_t` for endpoints in other nodes in the graph.

Correct me if i'm wrong but after I address https://github.com/ros2/rmw_zenoh/pull/171#discussion_r1648938166, it should be clear that the pointer is freed and that we pass a newly instantiated string to `TopicInfo`?

Reply:
> it's only unnecessary when the topics are created in the current session. However we still need to convert the string to a `rosidl_type_hash_t` for endpoints in other nodes in the graph since this information is gotten from liveliness tokens.

Oh yes, good point.  OK, never mind my comment on this.

> I was relying on [implicit conversion to a string](https://github.com/ros2/rmw_zenoh/blob/54c33bcd51381725ecb6e585db71679681ac08cf/rmw_zenoh_cpp/src/detail/liveliness_utils.hpp#L52) when passing `type_hash_c_str ` to the `TopicInfo` constructor. That should create a copy right?

Hm, I guess you are right.  Converting a `char *` to a `std::string` does indeed seem to do a copy of the string.  That's good to know.  Personally, I prefer this to be explicit, and so I would pass the `char *` into the `TopicInfo`, and have TopicInfo be responsible for both copying and releasing the string.  But I leave it to you.